### PR TITLE
Rsa key and new build option

### PR DESCRIPTION
**Pull request by joffreygonin**

### CODE REVIEW

**Comment by joffreygonin** [comment]
The second part of the build function (using the tags) does not work for me wet. You might want want me to fix this before accepting the pull request.

File: cloud_deploy/cli.py [hunk 1]
```diff
@@ -66,50 +66,86 @@ def cli(debug):
...
+        Build a Docker image locally and push to Docker Hub.
+        or 
+        Build the image in a remote server using --remote option
+    """
+
     repo = git.Repo('.', search_parent_directories=True)
     git_tag = repo.head.commit.hexsha[:7]
     if repo.is_dirty():
-        git_tag += "z"
+            git_tag += "z"
```
Comment (apdavison) [inline cloud_deploy/cli.py]: please remove extra space

File: cloud_deploy/cli.py [hunk 2]
```diff
@@ -66,50 +66,86 @@ def cli(debug):
...
-                   "date": datetime.now().isoformat()},
-                  fp)
-
-    click.echo("Building image")
-    result = shell.run(cmd.split(), cwd=build_directory, allow_error=True)
-    logger.debug(result.output)
-    if result.return_code != 0:
-        click.echo(result.output)
-        raise click.Abort()
+                "colour": colour,
```
Comment (apdavison) [inline cloud_deploy/cli.py]: please line-up the and following 2 lines with `"git"`

File: cloud_deploy/cli.py [hunk 3]
```diff
@@ -66,50 +66,86 @@ def cli(debug):
...
+                "date": datetime.now().isoformat()},
+                fp)
+
+    if remote == None :
```
Comment (apdavison) [inline cloud_deploy/cli.py]: Should be `if remote is None:`

File: cloud_deploy/cli.py [hunk 4]
```diff
@@ -66,50 +66,86 @@ def cli(debug):
...
+        result = shell.run(cmd.split(), cwd=build_directory, allow_error=True)
+
+        logger.debug(result.output)
+        if result.return_code != 0:
+            click.echo(result.output)
+            raise click.Abort()
+        
+    else :
+        #push project on remote
+        logger.info("Pushing project '{}' on the remote machine {} ".format(service, remote ))
```
Comment (apdavison) [inline cloud_deploy/cli.py]: should be "pushing ... to" not "pushing ... on"

File: cloud_deploy/cli.py [hunk 5]
```diff
@@ -66,50 +66,86 @@ def cli(debug):
...
-        shell.run(cmd.split())
-
-    # push image
-    cmd = "docker push {}:{}".format(image, colour_tag)
-    click.echo("Pushing image")
-    result = shell.run(cmd.split())
-    logger.debug(result.output)
-    logger.info("Pushed image {}:{}".format(image, colour_tag))
-
+        if remote == None :
```
Comment (apdavison) [inline cloud_deploy/cli.py]: `is None`

File: cloud_deploy/cli.py [hunk 6]
```diff
@@ -66,50 +66,86 @@ def cli(debug):
...
+            shell.run(cmd.split())
+        else : 
+            node._remote_execute(cmd)
+
+    if remote == None :
```
Comment (apdavison) [inline cloud_deploy/cli.py]: `is None`

File: cloud_deploy/nodes.py [hunk 7]
```diff
@@ -107,10 +107,10 @@ def create(cls, name, type="docker", size="512mb"):
                 region='ams2',
                 image=type,
                 size_slug=size,
-                ssh_keys=['66:0b:b5:20:a0:68:f9:fc:82:5a:de:c1:ce:03:4f:84'])
+                ssh_keys=['66:0b:b5:20:a0:68:f9:fc:82:5a:de:c1:ce:03:4f:84']) #think about what to do with that
         new_droplet.create()
         status = None
-        while status != "completed":
+        while status != "completed":
```
Comment (apdavison) [inline cloud_deploy/nodes.py]: please remote additional space at end of line

File: cloud_deploy/nodes.py [hunk 8]
```diff
@@ -107,10 +107,10 @@ def create(cls, name, type="docker", size="512mb"):
                 region='ams2',
                 image=type,
                 size_slug=size,
-                ssh_keys=['66:0b:b5:20:a0:68:f9:fc:82:5a:de:c1:ce:03:4f:84'])
+                ssh_keys=['66:0b:b5:20:a0:68:f9:fc:82:5a:de:c1:ce:03:4f:84']) #think about what to do with that
```
Comment (apdavison) [inline cloud_deploy/nodes.py]: Perhaps we should create a config file, where a list of ssh public keys could be included. In the meantime, perhaps just add your key fingerprint to this list
(run `ssh-keygen -E md5 -lf /path/to/your_key.pub` to generate the fingerprint)

File: cloud_deploy/nodes.py [hunk 9]
```diff
@@ -119,13 +119,27 @@ def create(cls, name, type="docker", size="512mb"):
...
-                    private_key_file=os.path.expanduser("~/.ssh/id_dsa"),  # to generalize - could be id_rsa, etc.
-                    missing_host_key=spur.ssh.MissingHostKey.warn)
-        with shell:
-            result = shell.run(shlex.split(cmd), cwd=cwd, encoding="utf-8")
-            return result.output
+        list_possible_keys_format = ["id_dsa", "id_rsa"]
+
+        #check if a corresponding key can be found
+        for key in list_possible_keys_format :            
+            if os.path.isfile(os.path.expanduser("~/.ssh/{}".format(key))) == False :
```
Comment (apdavison) [inline cloud_deploy/nodes.py]: `is False`

**Review by apdavison — CHANGES_REQUESTED** [review]
Nice work!

Just a handful of small formatting changes needed - see details

**Review by apdavison — APPROVED** [review]

**Pull request merged**